_posts/2019-01-21-api-best-practices.md

```diff
@@ -0,0 +1,90 @@
+type of error*](https://tools.ietf.org/html/rfc7231#section-6).  Use it!
+
+There's more to this than just being pedantic, by the way.  Since most developers will be interacting with HTTP
+through some kind of library (like Ruby's [RestClient](https://github.com/rest-client/rest-client), for instance),
+there's a good chance that the library provides support for fine grained error handling based on the specific error
+code being returned.  If you're not fully leveraging the HTTP status codes, then you're depriving your end users of
+that functionality.
+
+### Return structured error messages
+Speaking of errors - a status code is all well and good, if it's the right status code, but additional error messages
+are highly useful.  The HTTP standard allows you to return a body along with any error code, so be helpful and put
+additional information there. To really do it right, structure your error messages so that they can potentially
+contain more than just a brief message - a JSON encoded hash will work perfectly.  You can associate the error
+message with the same key every time, and include additional information as needed, like so:
+
+```
+{"msg":"User quota exceeded","max_allowed":10000,"administrator_id":1234}
+```
+
+In this example you not only get a message that tells you what went wrong (you're over quota), you're also told
+what your quota actually is, and the user id of the person to go bother about it.  Much more useful than just a
+status code or a generic error message alone.
+
+The gold standard for an API that does this well is the [GitHub API](https://developer.github.com/v3/).  Not only
+do you get a structured error message with a description, you also get *a link to the relevant documentation*.
+
+Which brings us to the next thing every good API needs:
+
+### Documentation, Documentation, Documentation
+
+All of the thoughtful design in the world won't mean a thing if no one knows how to use it.  Document every facet
+of your API - the individual endpoints, what they require, what is optional, and what they return, including errors
+and headers.  It's almost impossible to overdo this part.  If you find yourself wondering wether you should document
+some aspect of your API, the answer is "Yes."  Writing documentation is rarely fun, and writing it well is difficult.
+The difficulty is worth it, though, because the strength of documentation alone can often be a deciding factor in wether
+an API actually picks up users.
+
+These are just a few of the [many](https://blog.mwaysolutions.com/2014/06/05/10-best-practices-for-better-restful-api/), [many](https://github.com/RestCheatSheet/api-cheat-sheet#api-design-cheat-sheet), [many](https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api)
+different practices and design patterns that go into building an API that's a useful tool instead of an annoying chore.
+Make sure to implement these often neglected best practices, and you'll be well on your way.
```

Comment:
@rdormer This probably deserves its own paragraph.

Reply:
@etagwerker Right now it is it's own paragraph.  Are you saying it should be expanded some more?

Reply:
@rdormer I guess I got confused because I didn't see the empty line between the two sentences... If `Make sure to implement these often neglected best practices, and you'll be well on your way.` is a paragraph then we are good.